<h1>Hi-C library QC report</h1>
<p>&copy; 2018 Phase Genomics Inc.</p>
<h2>Library statistics</h2>
<table>
<thead>
<tr>
<th>Label</th>
<th>Your library</th>
<th>Expected values</th>
</tr>
</thead>
<tbody>
<tr>
<td>BAM file</td>
<td>/Users/maximilianpress/Downloads/abc_qc_control1.bam</td>
<td>N/A</td>
</tr>
<tr>
<td>Number of read pairs</td>
<td>414177</td>
<td>N/A</td>
</tr>
<tr>
<td>Fraction of read pairs &gt;10KB apart</td>
<td>0.009</td>
<td>0.005-0.05</td>
</tr>
<tr>
<td>Fraction of read pairs mapping to different contigs/chromosomes</td>
<td>0.236</td>
<td>0.1-0.5 (contigs) 0.01-0.1 (chromosomes)</td>
</tr>
<tr>
<td>Fraction of split reads</td>
<td>0.267</td>
<td>0.1-0.4 (PG libraries) 0.3+ (other libraries)</td>
</tr>
<tr>
<td>Fraction of zero-distance pairs</td>
<td>0.257</td>
<td>0-0.15</td>
</tr>
<tr>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>See below for information on differences between Phase Genomics Hi-C libraries and traditional Hi-C libraries.</p>
<h2>Aligned mate distance histograms</h2>
<p><img alt="&quot;Long range interaction histogram&quot;" src="/Users/maximilianpress/PGTools/bam_to_mate_hist/Read_mate_dist_long.png" />
<img alt="&quot;Short range interaction histogram&quot;" src="/Users/maximilianpress/PGTools/bam_to_mate_hist/Read_mate_dist_short.png" /></p>
<h2>Alignment distance statistics and plots</h2>
<p>We briefly describe some of the statistics we compute below to aid interpretation of this report.</p>
<h3>Fraction of read pairs &gt; 10KB apart</h3>
<h3>Fraction of read pairs mapping to different contigs or chromosomes</h3>
<h3>Fraction of zero-distance pairs</h3>
<h2>Split reads</h2>
<p>Traditionally, split reads have been a favored measure of Hi-C library quality. This is because traditional Hi-C library preparations are expected to produce many reads reading through junctions. </p>
<p>Phase Genomics libraries, whether produced in our laboratory or by means of ProxiMeta &copy;, Animal, Plant, or Human kits, will have a generally lower fraction of split reads. This is because we have optimized our Hi-C protocol to </p>